src/main/groovy/com/twcable/grabbit/jcr/ACLProtoNodeDecorator.groovy

```diff
@@ -0,0 +1,240 @@
+        this.nameOverride = nameOverride
+    }
+
+
+    @Override
+    JCRNodeDecorator writeToJcr(@Nonnull Session session) {
+        /**
+         * We don't write the rep:policy node directly. Rather, we find the rep:policy node's ACE(s) and add them to the
+         * owner's existing policy; or we add them to a new policy.
+         */
+        clearExistingEntries(session)
+        innerProtoNode.mandatoryChildNodeList.each { ProtoNode node ->
+            if(isGrantACEType(node)) {
+                writeGrantACE(session, node)
+            }
+            else if(isDenyACEType(node)) {
+                writeDenyACE(session, node)
+            }
+        }
+        session.save()
+        //rep:policy should exist now, and should be free to return within a JCRNodeDecorator
+        return new JCRNodeDecorator(session.getNode(getName()))
+    }
+
+
+    private void clearExistingEntries(final Session session) {
+        final JackrabbitAccessControlList acl = getAccessControlList(session)
+        if(acl.accessControlEntries.length != 0) {
+            acl.accessControlEntries.each { AccessControlEntry entry ->
+                acl.removeAccessControlEntry(entry)
+            }
+            getAccessControlManager(session).setPolicy(getParentPath(), acl)
+            session.save()
+        }
+    }
+
+
+    private void writeGrantACE(final Session session, ProtoNode grantACENode) {
+        writeACE(session, grantACENode, true)
+    }
+
+
+    private void writeDenyACE(final Session session, ProtoNode denyACENode) {
+        writeACE(session, denyACENode, false)
+    }
+
+
+    private void writeACE(final Session session, ProtoNode aceNode, boolean grant) {
+        JackrabbitAccessControlList acl = getAccessControlList(session)
+        Principal principal = getPrincipal(session, getPrincipalName(aceNode))
+        Privilege[] privileges = getPrivilegeNames(aceNode).collect { String privilegeName ->
+            if(isSupportedPrivilege(session, privilegeName)) {
+                return getPrivilege(session, privilegeName)
+            }
+            else {
+                throw new IllegalStateException("${privilegeName} is not a supported privilege on this JCR implementation. Bailing out")
+            }
+        }
+        Map<String, Value> restrictionMap
+        if(hasRestrictions(aceNode)) {
+            ProtoNode restrictionNode = getRestrictionNode(aceNode)
+            restrictionMap = extractRestrictionMapFrom(restrictionNode)
+        }
+        if(restrictionMap != null) {
+            acl.addEntry(principal, privileges, grant, restrictionMap)
+        }
+        else {
+            acl.addEntry(principal, privileges, grant)
+        }
+        getAccessControlManager(session).setPolicy(getParentPath(), acl)
+    }
+
+
+    private boolean isGrantACEType(ProtoNode node) {
+        return node.propertiesList.collect { new ProtoPropertyDecorator(it) }.find { ProtoPropertyDecorator property ->
```

Comment:
Yeah, I think we can wrap, and call isGrantACEType in the same iteration. Makes sense